fleetspeak/src/server/comms.go

```diff
@@ -46,9 +46,113 @@ type commsContext struct {
 	s *Server
 }
 
-// GetClientInfo loads basic information about a client. Returns nil if the client does
+func (c commsContext) InitializeConnection(ctx context.Context, addr net.Addr, key crypto.PublicKey, wcd *fspb.WrappedContactData) (*comms.ConnectionInfo, *fspb.ContactData, error) {
+	id, err := common.MakeClientID(key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	contactInfo := authorizer.ContactInfo{
+		ID:           id,
+		ContactSize:  len(wcd.ContactData),
+		ClientLabels: wcd.ClientLabels,
+	}
+	if !c.s.authorizer.Allow2(addr, contactInfo) {
+		return nil, nil, comms.NotAuthorizedError
+	}
+
+	ci, err := c.getClientInfo(ctx, id)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	res := comms.ConnectionInfo{
+		Addr: addr,
+	}
+	if ci == nil {
+		res.AuthClientInfo.New = true
+	} else {
+		res.Client = *ci
+		res.AuthClientInfo.Labels = ci.Labels
+	}
+
+	if !c.s.authorizer.Allow3(addr, contactInfo, res.AuthClientInfo) {
+		return nil, nil, comms.NotAuthorizedError
+	}
+
+	if ci == nil {
+		if err := c.addClient(ctx, id, key); err != nil {
+			return nil, nil, err
+		}
+		res.Client.ID = id
+		res.Client.Key = key
+		// Set initial labels for the client according to the contact data. Going
+		// forward, labels will be adjusted when the client sends a ClientInfo
+		// message (see system_service.go).
+		for _, l := range wcd.ClientLabels {
+			cl := &fspb.Label{ServiceName: "client", Label: l}
+
+			// Ignore errors - if this fails, the first ClientInfo message will try again
+			// in a context where we can retry easily.
+			c.s.dataStore.AddClientLabel(ctx, id, cl)
+
+			res.Client.Labels = append(res.Client.Labels, cl)
+		}
+		res.AuthClientInfo.Labels = res.Client.Labels
+	}
+
+	sigs, err := signatures.ValidateWrappedContactData(id, wcd)
+	if err != nil {
+		return nil, nil, err
+	}
+	accept, validationInfo := c.s.authorizer.Allow4(
+		addr,
+		contactInfo,
+		res.AuthClientInfo,
+		sigs)
+	if !accept {
+		return nil, nil, errors.New("contact not authorized")
```

Comment:
Any reason why you cannot return the comms.NotAuthorizedError here?

Reply:
Just a failed search and replace. Fixing...